File: 1846-maximum-element-after-decreasing-and-rearranging/1846-maximum-element-after-decreasing-and-rearranging.py
class Solution:
    def maximumElementAfterDecrementingAndRearranging(self, arr: List[int]) -> int:

        # Sort the input array in ascending order
        arr.sort()

        # Ensure the first element is 1
        if arr[0] != 1:
            arr[0] = 1

        # Adjust values to minimize differences between adjacent elements
        for i in range(1, len(arr)):
            # Calculate the absolute difference between the current and previous elements
            difference = abs(arr[i] - arr[i - 1])

            # Check if the difference is less than or equal to 1
            if difference <= 1:
                # If the difference is within the allowed range, no adjustment is needed
                pass
            else:
                # If the difference is greater than 1, adjust the value to minimize differences
                temp = arr.pop(i)
                arr.insert(i, temp - difference + 1)

        # Return the maximum element in the modified array
        return arr[-1]

File: 1846-maximum-element-after-decreasing-and-rearranging/1846-maximum-element-after-decreasing-and-rearranging.java
class Solution {
    public int maximumElementAfterDecrementingAndRearranging(int[] arr) {
        
        // Step 1: Sort the array in ascending order
        Arrays.sort(arr);
        
        // Step 2: Ensure the first element is 1, as it needs to be the minimum possible value
        if (arr[0] != 1) {
            arr[0] = 1;
        }
        
        // Step 3: Iterate through the array starting from the second element
        for (int i = 1; i < arr.length; i++){
            
            // Calculate the absolute difference between the current element and the previous one
            int difference = Math.abs(arr[i] - arr[i - 1]);
            
            // Step 4: If the difference is less than or equal to 1, no adjustment is needed
            if (difference <= 1){
                // Do nothing in this case
            } else {
                // Step 5: Adjust the current element to make the difference exactly 1
                // by adding 1 and subtracting the calculated difference
                arr[i] = arr[i] + 1 - difference;
            }
        }
        
        // Step 6: Return the maximum element in the modified array
        return arr[arr.length - 1];
    }
}
